mne/cov.py

```diff
@@ -876,67 +937,83 @@ def _check_scalings_user(scalings):
 
 
 def _compute_covariance_auto(data, method, info, method_params, cv,
-                             scalings, n_jobs, stop_early, picks_list):
+                             scalings, n_jobs, stop_early, picks_list, rank):
     """Compute covariance auto mode."""
     # rescale to improve numerical stability
     _apply_scaling_array(data.T, picks_list=picks_list, scalings=scalings)
+    if rank != 'full':
+        C = np.dot(data.T, data)
+        eig, eigvec = _smart_eigh(C, info, rank, scalings=1.)  # already scaled
+        mask = eig > 0
+        del eig
+        eigvec = eigvec[mask]
+        data = np.dot(data, eigvec.T)
+        used = np.where(mask)[0]
+        sub_picks_list = [(key, np.searchsorted(used, picks))
+                          for key, picks in picks_list]
+        sub_info = pick_info(info, used)
+        logger.info('Reducing data rank from %s -> %s'
+                    % (len(mask), eigvec.shape[0]))
+    else:
+        eigvec = None
+        sub_picks_list = picks_list
+        sub_info = info
     estimator_cov_info = list()
     msg = 'Estimating covariance using %s'
 
     ok_sklearn = check_version('sklearn', '0.15')
     if not ok_sklearn and (len(method) != 1 or method[0] != 'empirical'):
         raise ValueError('scikit-learn is not installed, `method` must be '
-                         '`empirical`')
+                         '`empirical`, got %s' % (method,))
 
     for this_method in method:
         data_ = data.copy()
         name = this_method.__name__ if callable(this_method) else this_method
         logger.info(msg % name.upper())
         mp = method_params[this_method]
-        _info = None
 
         if this_method == 'empirical':
             est = EmpiricalCovariance(**mp)
             est.fit(data_)
-            estimator_cov_info.append((est, est.covariance_, _info))
+            estimator_cov_info.append([est, est.covariance_, None])
```

Comment:
why do we convert those to lists?

Reply:
Might not be necessary (can't remember if I needed it at one point), but I also don't mind it because visually it helps separate the append method parens from what is being appended

Reply:
Ahh yes it is necessary:
```
        # undo scaling
        for c in estimator_cov_info:
            if eigvec is not None:
                # project back if necessary
>               c[1] = np.dot(eigvec.T, np.dot(c[1], eigvec))
```

Reply:
Ok cool. Just wanted to make sure we have some reasons.